### ISSUE TITLE
TimeoutPersisterReceiver: The maximum number of requests (30) allowed for this session has been reached.

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.RavenDB#246]
## Who's affected
Users with a very large number of timeouts (in excess of 30,720) that have expired.

### 3.x

 https://github.com/Particular/NServiceBus.RavenDB/blob/support-3.0/src/NServiceBus.RavenDB/Timeouts/TimeoutPersister.cs#L42

### 1.x

https://github.com/Particular/NServiceBus.RavenDB/blob/support-1.0/src/NServiceBus.RavenDB/TimeoutPersister/RavenTimeoutPersistence.cs#L105

## Symptoms
The TimeoutPersisterReceiver fails with the following exception:

```
WARN  NServiceBus.Timeout.Hosting.Windows.TimeoutPersisterReceiver Failed to fetch timeouts from the timeout storage
System.InvalidOperationException: The maximum number of requests (30) allowed for this session has been reached.
Raven limits the number of remote calls that a session is allowed to make as an early warning system. Sessions are expected to be short lived, and 
Raven provides facilities like Load(string[] keys) to load multiple documents at once and batch saves (call SaveChanges() only once).
You can increase the limit by setting DocumentConvention.MaxNumberOfRequestsPerSession or MaxNumberOfRequestsPerSession, but it is
advisable that you'll look into reducing the number of remote calls first, since that will speed up your application significantly and result in a 
more responsive application.

   at Raven.Client.Document.InMemoryDocumentSessionOperations.IncrementRequestCount() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\InMemoryDocumentSessionOperations.cs:line 349
   at Raven.Client.Document.AbstractDocumentQuery`2.ExecuteActualQuery() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\AbstractDocumentQuery.cs:line 679
   at Raven.Client.Document.AbstractDocumentQuery`2.InitSync() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\AbstractDocumentQuery.cs:line 666
   at Raven.Client.Document.AbstractDocumentQuery`2.get_QueryResult() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Document\AbstractDocumentQuery.cs:line 653
   at Raven.Client.Linq.RavenQueryProviderProcessor`1.ExecuteQuery[TProjection]() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Linq\RavenQueryProviderProcessor.cs:line 1628
   at Raven.Client.Linq.RavenQueryProviderProcessor`1.Execute(Expression expression) in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Linq\RavenQueryProviderProcessor.cs:line 1600
   at Raven.Client.Linq.RavenQueryProvider`1.Execute(Expression expression) in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Linq\RavenQueryProvider.cs:line 154
   at Raven.Client.Linq.RavenQueryProvider`1.System.Linq.IQueryProvider.Execute(Expression expression) in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Linq\RavenQueryProvider.cs:line 212
   at Raven.Client.Linq.RavenQueryInspector`1.GetEnumerator() in c:\Builds\RavenDB-Stable-3.0\Raven.Client.Lightweight\Linq\RavenQueryInspector.cs:line 101
   at NServiceBus.TimeoutPersisters.RavenDB.TimeoutPersister.GetNextChunk(DateTime startSlice, DateTime& nextTimeToRunQuery) in C:\BuildAgent\work\c4d62ce02b983878\src\NServiceBus.RavenDB\Timeouts\TimeoutPersister.cs:line 71
   at NServiceBus.Timeout.Hosting.Windows.TimeoutPersisterReceiver.Poll(Object obj) in C:\BuildAgent\work\3206e2123f54fce4\src\NServiceBus.Core\Timeout\Hosting\Windows\TimeoutPersisterReceiver.cs:line 90
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
```

The timeout persister does a while loop until all timeouts are fetched. If a lot of timeouts are expired this will result in MaxNumberOfRequestsPerSession to exceed its tresshold and RavenDB to throw an InvalidOperationException

Case: https://nservicebus.desk.com/agent/case/18477

**Comment by ramonsmits** [comment]
This bug is introduced due an incorrect solution in #140

**Comment by mauroservienti** [comment]
v3.x is now fixed in v3.0.11, I'll assign this one to the v4 milestone to track that we need to apply the same fix to vNext

**Comment by mauroservienti** [comment]
Should we create a separate issue for vNext so that this can be included in 3.0.11 release notes?

**Comment by DavidBoike** [comment]
Yes. I organized the PR #249 so that it could be the release notes bearer for 3.0.11 but then I wasn't happy with the release, so it would be easier to bring that content in here and have a new issue for vNext.

**Comment by mauroservienti** [comment]
Since the PR is merged can we close this as fixed?

**Comment by mauroservienti** [comment]
In support-3.0 the [`TimeoutPersister`](https://github.com/Particular/NServiceBus.RavenDB/blob/support-3.0/src/NServiceBus.RavenDB/Timeouts/TimeoutPersister.cs#L68) logs the number of timeouts that will be returned, in hotfix-3.0.1 the logs are gone, should we bring them back? and port them to [v4](https://github.com/Particular/NServiceBus.RavenDB/issues/247) as well?

**Comment by DavidBoike** [comment]
@mauroservienti I don't think so. It was just noisy, and didn't take that long even for a large number of timeouts.

**Comment by mauroservienti** [comment]
So is this done and is just waiting for the Hotfix release?

**Comment by DavidBoike** [comment]
@ramonsmits Sorry for the delay. The Desk case link is of course dead. Can you find the case back and follow up with the customer?

**Comment by ramonsmits** [comment]
@DavidBoike I'll try....
